BrainPortal/cbrain_plugins/cbrain-plugins-base/cbrain_task/start_vm/bourreau/start_vm.rb

```diff
@@ -0,0 +1,242 @@
+
+#
+# CBRAIN Project
+#
+# Copyright (C) 2008-2012
+# The Royal Institution for the Advancement of Learning
+# McGill University
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.  
+#
+
+require 'net/ssh'
+
+# A task that starts a VM, monitors its booting process using ssh, and
+# mounts the Bourreau task and data directories in it using sshfs.
+class CbrainTask::StartVM < ClusterTask
+
+  Revision_info = CbrainFileRevision[__FILE__]
+
+  # A Hook used to follow the boot process of the VM once it is
+  # running.
+  after_status_transition '*', 'On CPU', :starting
+
+  def setup #:nodoc:
+    validate_params # Defined in common, will raise an exception if
+                    # params aren't valid.
+    true
+  end
+  
+  def cluster_commands #:nodoc:
+    [ "echo This will never execute" ]  # If the cluster commands are
+    # empty, task will jump directly
+    # to state data ready.
+  end
+  
+  def save_results #:nodoc:
+    addlog("No result to save.")
+    # We consider the task successful if the VM booted. 
+    return true if params[:vm_status] == "booted"
+    addlog "VM is not active and never booted. I don't know why, sorry."
+    return false
+  end
+
+  # Hook called when the VM task gets on CPU. 
+  # Monitors the VM until it boots. Then mounts shared directories with sshfs. 
+  def starting(init_status) #:nodoc:
+    params = self.params
+
+    # Updates VM local IP now that it is running.
+    params[:vm_local_ip] = scir_session.get_local_ip(id)
+    if params[:vm_local_ip].blank? 
+      addlog "Cannot get VM local IP. Terminating VM task (id = #{id})."
+      update_vm_status("unreachable")
+      save!
+      terminate
+    end
+    update_vm_status("booting")
+    save! 
+
+    # Monitors the VM until it has booted.
+    CBRAIN.spawn_with_active_records("Monitor VM")  do
```

Comment:
This will start a BG process on the Bourreau front end... how many of these can there be? Can it go to the hundreds?!?